In [260]:
import kagglehub
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import xgboost as xgb
import gpboost as gpb
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score, RandomizedSearchCV
from sklearn.metrics import mean_squared_error

# Gathering the data

In [261]:
path = kagglehub.dataset_download("nishaanamin/march-madness-data")
print("Path to dataset files:", path)

Path to dataset files: /Users/kolbytaylor/.cache/kagglehub/datasets/nishaanamin/march-madness-data/versions/100


In [262]:
# Loading all the different data sets in

# Tournament Results and Data
coach_results = pd.read_csv("march-madness-data/versions/83/Coach Results.csv")
seed_results = pd.read_csv("march-madness-data/versions/83/Seed Results.csv")
team_results = pd.read_csv("march-madness-data/versions/83/Team Results.csv")
conference_results = pd.read_csv("march-madness-data/versions/83/Conference Results.csv")
tournament_locations = pd.read_csv("march-madness-data/versions/83/Tournament Locations.csv")
tournament_matchups = pd.read_csv("march-madness-data/versions/83/Tournament Matchups.csv")
upset_counts = pd.read_csv("march-madness-data/versions/83/Upset Count.csv")
upset_seed_info = pd.read_csv("march-madness-data/versions/83/Upset Seed Info.csv")

# Bartorvik
barttorvik_away_neutral = pd.read_csv("march-madness-data/versions/83/Barttorvik Away-Neutral.csv")
barttorvik_away = pd.read_csv("march-madness-data/versions/83/Barttorvik Away.csv")
barttorvik_home = pd.read_csv("march-madness-data/versions/83/Barttorvik Home.csv")
barttorvik_neutral = pd.read_csv("march-madness-data/versions/83/Barttorvik Home.csv")
kenpom_barttorvik = pd.read_csv("march-madness-data/versions/83/KenPom Barttorvik.csv")

# Conference Stats
conference_stats = pd.read_csv("march-madness-data/versions/83/Conference Stats.csv")
conference_away_neutral = pd.read_csv("march-madness-data/versions/83/Conference Stats Away Neutral.csv")
conference_away = pd.read_csv("march-madness-data/versions/83/Conference Stats Away.csv")
conference_home = pd.read_csv("march-madness-data/versions/83/Conference Stats Home.csv")
conference_neutral = pd.read_csv("march-madness-data/versions/83/Conference Stats Neutral.csv")

# Resumes
resumes = pd.read_csv("march-madness-data/versions/83/Resumes.csv")

# Shooting Splits
shooting_splits = pd.read_csv("march-madness-data/versions/83/Shooting Splits.csv")

# Other Websites
ratings_538 = pd.read_csv("march-madness-data/versions/83/538 Ratings.csv")
heat_check = pd.read_csv("march-madness-data/versions/83/Heat Check Tournament Index.csv")

# People's Opinions
preseason_votes = pd.read_csv("march-madness-data/versions/83/Preseason Votes.csv")
public_picks = pd.read_csv("march-madness-data/versions/83/Public Picks.csv")

In [263]:
# Take the columns in the dataframes that have home, away, and neutral; and subscript them so that we can recognize what it is
# Barttorvik
for col in barttorvik_away_neutral.columns:
    barttorvik_away_neutral = barttorvik_away_neutral.rename(columns={col: col + " AWAY-NEUTRAL"})

for col in barttorvik_away.columns:
    barttorvik_away = barttorvik_away.rename(columns={col: col + " AWAY"})

for col in barttorvik_home.columns:
    barttorvik_home = barttorvik_home.rename(columns={col: col + " HOME"})

for col in barttorvik_neutral.columns:
    barttorvik_neutral = barttorvik_neutral.rename(columns={col: col + " NEUTRAL"})



# Conference Statistics
for col in conference_away_neutral.columns:
    conference_away_neutral = conference_away_neutral.rename(columns={col: col + " AWAY-NEUTRAL"})

for col in conference_away.columns:
    conference_away = conference_away.rename(columns={col: col + " AWAY"})

for col in conference_home.columns:
    conference_home = conference_home.rename(columns={col: col + " HOME"})

for col in conference_neutral.columns:
    conference_neutral = conference_neutral.rename(columns={col: col + " NEUTRAL"})


In [264]:
barttorvik_away_neutral.head()

YEAR AWAY-NEUTRAL  TEAM NO AWAY-NEUTRAL  TEAM ID AWAY-NEUTRAL  \
0               2024                  1079                   197   
1               2024                  1078                    48   
2               2024                  1077                    33   
3               2024                  1076                    43   
4               2024                  1075                    36   

  TEAM AWAY-NEUTRAL  SEED AWAY-NEUTRAL  ROUND AWAY-NEUTRAL  \
0             Akron                 14                   0   
1           Alabama                  4                   0   
2           Arizona                  2                   0   
3            Auburn                  4                   0   
4            Baylor                  3                   0   

   BADJ EM AWAY-NEUTRAL  BADJ O AWAY-NEUTRAL  BADJ D AWAY-NEUTRAL  \
0                   3.1                105.0                101.9   
1                  18.7                125.4                106.7   
2                  23.7                119.7                 96.0   
3                  26.8                120.8                 94.0   
4                  17.6                117.7                100.1   

   BARTHAG AWAY-NEUTRAL  ...  BADJT RANK AWAY-NEUTRAL  \
0                 0.585  ...                      265   
1                 0.865  ...                       10   
2                 0.927  ...                       40   
3                 0.947  ...                       67   
4                 0.866  ...                      275   

   AVG HGT RANK AWAY-NEUTRAL  EFF HGT RANK AWAY-NEUTRAL  \
0                        238                        199   
1                         33                          8   
2                         50                         37   
3                         86                         76   
4                         31                         22   

   EXP RANK AWAY-NEUTRAL  TALENT RANK AWAY-NEUTRAL  FT% RANK AWAY-NEUTRAL  \
0                     19                       176                    186   
1                    156                       106                     12   
2                    196                         7                    157   
3                    127                        69                    127   
4                    304                        34                    137   

   OP FT% RANK AWAY-NEUTRAL  PPPO RANK AWAY-NEUTRAL  PPPD RANK AWAY-NEUTRAL  \
0                       188                     120                      47   
1                       357                       3                     351   
2                       284                      19                      98   
3                       260                       5                      18   
4                       248                      52                     189   

   ELITE SOS RANK AWAY-NEUTRAL  
0                          269  
1                            8  
2                           52  
3                           77  
4                           11  

[5 rows x 85 columns]

There are many columns in the Barttorvik data, and splitting them out by where the team played multiplies it by five. In total we have around 170 features. Half of them are metrics and the other half rank those metrics in all Division 1. We should probably sort these out first. I'm going to try doing some analysis with the features. I'm going to fit the same model with the overall, away, netural, home, and neutral-away features and see which ones work the best, as well as do some AB testing to see if there's a significant difference between them.

In [265]:
# Creates the target, 0-6 based on how many games were won
conditions = [
    (kenpom_barttorvik['ROUND'] == 0) | (kenpom_barttorvik['ROUND'] == 68),
    (kenpom_barttorvik['ROUND'] == 64),
    (kenpom_barttorvik['ROUND'] == 32),
    (kenpom_barttorvik['ROUND'] == 16),
    (kenpom_barttorvik['ROUND'] == 8),
    (kenpom_barttorvik['ROUND'] == 4),
    (kenpom_barttorvik['ROUND'] == 2),
    (kenpom_barttorvik['ROUND'] == 1)
]

values = [np.nan, 0, 1, 2, 3, 4, 5, 6]

kenpom_barttorvik['TARGET'] = np.select(conditions, values, default=np.nan)
barttorvik_away_neutral['TARGET'] = np.select(conditions, values, default=np.nan)
barttorvik_away['TARGET'] = np.select(conditions, values, default=np.nan)
barttorvik_home['TARGET'] = np.select(conditions, values, default=np.nan)
barttorvik_neutral['TARGET'] = np.select(conditions, values, default=np.nan)

In [266]:
barttorvik_away_neutral.head()

YEAR AWAY-NEUTRAL  TEAM NO AWAY-NEUTRAL  TEAM ID AWAY-NEUTRAL  \
0               2024                  1079                   197   
1               2024                  1078                    48   
2               2024                  1077                    33   
3               2024                  1076                    43   
4               2024                  1075                    36   

  TEAM AWAY-NEUTRAL  SEED AWAY-NEUTRAL  ROUND AWAY-NEUTRAL  \
0             Akron                 14                   0   
1           Alabama                  4                   0   
2           Arizona                  2                   0   
3            Auburn                  4                   0   
4            Baylor                  3                   0   

   BADJ EM AWAY-NEUTRAL  BADJ O AWAY-NEUTRAL  BADJ D AWAY-NEUTRAL  \
0                   3.1                105.0                101.9   
1                  18.7                125.4                106.7   
2                  23.7                119.7                 96.0   
3                  26.8                120.8                 94.0   
4                  17.6                117.7                100.1   

   BARTHAG AWAY-NEUTRAL  ...  AVG HGT RANK AWAY-NEUTRAL  \
0                 0.585  ...                        238   
1                 0.865  ...                         33   
2                 0.927  ...                         50   
3                 0.947  ...                         86   
4                 0.866  ...                         31   

   EFF HGT RANK AWAY-NEUTRAL  EXP RANK AWAY-NEUTRAL  TALENT RANK AWAY-NEUTRAL  \
0                        199                     19                       176   
1                          8                    156                       106   
2                         37                    196                         7   
3                         76                    127                        69   
4                         22                    304                        34   

   FT% RANK AWAY-NEUTRAL  OP FT% RANK AWAY-NEUTRAL  PPPO RANK AWAY-NEUTRAL  \
0                    186                       188                     120   
1                     12                       357                       3   
2                    157                       284                      19   
3                    127                       260                       5   
4                    137                       248                      52   

   PPPD RANK AWAY-NEUTRAL  ELITE SOS RANK AWAY-NEUTRAL  TARGET  
0                      47                          269     NaN  
1                     351                            8     NaN  
2                      98                           52     NaN  
3                      18                           77     NaN  
4                     189                           11     NaN  

[5 rows x 86 columns]

In [267]:
X_barttorvik = kenpom_barttorvik.drop(columns=['YEAR', 'CONF', 'CONF ID', 'QUAD NO', 'QUAD ID', 'TEAM NO', 'TEAM ID', 'TEAM', 'SEED', 'ROUND', 'TARGET'])
y_barttorvik = kenpom_barttorvik['TARGET']

X_barttorvik_away_neutral = barttorvik_away_neutral.drop(columns=['YEAR AWAY-NEUTRAL', 'TEAM NO AWAY-NEUTRAL', 'TEAM ID AWAY-NEUTRAL', 'TEAM AWAY-NEUTRAL', 'SEED AWAY-NEUTRAL', 'ROUND AWAY-NEUTRAL', 'TARGET'])
y_barttorvik_away_neutral = barttorvik_away_neutral['TARGET']

X_barttorvik_away = barttorvik_away.drop(columns=['YEAR AWAY', 'TEAM NO AWAY', 'TEAM ID AWAY', 'TEAM AWAY', 'SEED AWAY', 'ROUND AWAY', 'TARGET'])
y_barttorvik_away = barttorvik_away['TARGET']

X_barttorvik_home = barttorvik_home.drop(columns=['YEAR HOME', 'TEAM NO HOME', 'TEAM ID HOME', 'TEAM HOME', 'SEED HOME', 'ROUND HOME', 'TARGET'])
y_barttorvik_home = barttorvik_home['TARGET']

X_barttorvik_neutral = barttorvik_neutral.drop(columns=['YEAR NEUTRAL', 'TEAM NO NEUTRAL', 'TEAM ID NEUTRAL', 'TEAM NEUTRAL', 'SEED NEUTRAL', 'ROUND NEUTRAL', 'TARGET'])
y_barttorvik_neutral = barttorvik_neutral['TARGET']

# Drop rows where TARGET is null
X_barttorvik = X_barttorvik.loc[y_barttorvik.notna()]
y_barttorvik = y_barttorvik.dropna()

X_barttorvik_away_neutral = X_barttorvik_away_neutral.loc[y_barttorvik_away_neutral.notna()]
y_barttorvik_away_neutral = y_barttorvik_away_neutral.dropna()

X_barttorvik_away = X_barttorvik_away.loc[y_barttorvik_away.notna()]
y_barttorvik_away = y_barttorvik_away.dropna()

X_barttorvik_home = X_barttorvik_home.loc[y_barttorvik_home.notna()]
y_barttorvik_home = y_barttorvik_home.dropna()

X_barttorvik_neutral = X_barttorvik_neutral.loc[y_barttorvik_neutral.notna()]
y_barttorvik_neutral = y_barttorvik_neutral.dropna()


I don't want to have to do feature scaling or do anything with assumptions because this is just EDA modeling, so I'm going to use XGBoost.

In [268]:
# No null values! That's nice. 
print(X_barttorvik.isnull().sum()[X_barttorvik.isnull().sum() > 0])

Series([], dtype: int64)


In [269]:
# Datatypes look good
print(X_barttorvik.dtypes)

K TEMPO           float64
K TEMPO RANK        int64
KADJ T            float64
KADJ T RANK         int64
K OFF             float64
                   ...   
FT% RANK            int64
OP FT% RANK         int64
PPPO RANK           int64
PPPD RANK           int64
ELITE SOS RANK      int64
Length: 93, dtype: object


In [270]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik, y_barttorvik, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24614
[10]	eval-rmse:1.07732
[20]	eval-rmse:1.08317
[30]	eval-rmse:1.08900
[33]	eval-rmse:1.09478
Best iteration: 14
Test RMSE: 1.0766


In [271]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik_away_neutral, y_barttorvik_away_neutral, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24819
[10]	eval-rmse:1.08775
[20]	eval-rmse:1.08954
[30]	eval-rmse:1.09811
[33]	eval-rmse:1.10664
Best iteration: 13
Test RMSE: 1.0897


In [272]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik_away, y_barttorvik_away, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24884
[10]	eval-rmse:1.12063
[20]	eval-rmse:1.11433
[30]	eval-rmse:1.12300
[33]	eval-rmse:1.13158
Best iteration: 14
Test RMSE: 1.1112


In [273]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik_home, y_barttorvik_home, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24086
[10]	eval-rmse:1.07804
[20]	eval-rmse:1.08300
[30]	eval-rmse:1.09856
Best iteration: 10
Test RMSE: 1.0840


In [274]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik_neutral, y_barttorvik_neutral, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24086
[10]	eval-rmse:1.07804
[20]	eval-rmse:1.08300
[29]	eval-rmse:1.09671
Best iteration: 10
Test RMSE: 1.0840


### Model Results:

The results were consistent across all models. The hyperparameters selected by each model were identical:

- subsample: 0.6  
- min_child_weight: 5  
- max_depth: 2  
- learning_rate: 0.05  
- lambda: 5  
- colsample_bytree: 1.0  
- alpha: 10  

---

### RMSE:

- Overall: 1.07  
- Away-Neutral: 1.08  
- Away: 1.11  
- Home: 1.08  
- Neutral: 1.08  

---

The Overall RMSE might be slightly better due to the inclusion of a few different features. Based on this analysis, we have decided to use the overall Barttorvik data in our analysis moving forward.


In [275]:
# Starting with the Kenpom data
all_data = kenpom_barttorvik

# Add 538 Ratings
all_data = all_data.merge(ratings_538, how='left', on='TEAM NO')
all_data = all_data.drop(columns=['YEAR_y', 'TEAM_y', 'SEED_y', 'ROUND_y'])
all_data = all_data.rename(columns={'YEAR_x': 'YEAR', 'TEAM_x': 'TEAM', 'SEED_x': 'SEED', 'ROUND_x': 'ROUND'})

# Add Heat Check metrics
all_data = all_data.merge(heat_check, how='left', on='TEAM NO')
all_data = all_data.drop(columns=['YEAR_y', 'TEAM_y', 'SEED_y', 'ROUND_y'])
all_data = all_data.rename(columns={'YEAR_x': 'YEAR', 'TEAM_x': 'TEAM', 'SEED_x': 'SEED', 'ROUND_x': 'ROUND'})

# Add resume data, includes strength of schedule
all_data = all_data.merge(resumes, how='left', on='TEAM NO')
all_data = all_data.drop(columns=['YEAR_y', 'TEAM_y'])
all_data = all_data.rename(columns={'YEAR_x': 'YEAR', 'TEAM_x': 'TEAM', 'SEED_x': 'SEED', 'ROUND_x': 'ROUND'})

# Shooting splits of each team
all_data = all_data.merge(shooting_splits, how='left', on='TEAM NO')
all_data = all_data.drop(columns=['YEAR_y', 'TEAM_y', 'SEED_y', 'ROUND_y'])
all_data = all_data.rename(columns={'YEAR_x': 'YEAR', 'TEAM_x': 'TEAM', 'SEED_x': 'SEED', 'ROUND_x': 'ROUND', 'CONF_x': 'CONF', 'TEAM ID_x': 'TEAM ID'})

# Drops any more columns that don't seem useful
all_data = all_data.drop(columns=['CONF ID', 'QUAD NO', 'QUAD ID', 'TEAM NO', 'TEAM ID', 'TARGET', 'VAL Z-SCORE'])

There are coach and conference results that seem cool and would be interesting, but scare me because of potential data leakage. For example, Bill Self won a championship a few years ago, but before that his championships would be at zero... So I don't think it's a good idea to include it. Conference results are similar, but might be more watered down because they're average over the entire conference.

In [276]:
# Prints the number of columns
print(len(all_data.columns))

154


Our feature space is really high, so we're going to try and reduce it by removing the columns that rank the features. These are obviously highly correlated with the features themselves, we're just losing some information about how the team compared to other teams in its season.

In [277]:
all_data = all_data.drop(columns=[col for col in all_data.columns if 'RANK' in col])

In [278]:
# Prints the number of columns
print(len(all_data.columns))

90


Okay, now we're down to a more reasonable space. I'm sure there's some unimportant features, or features with high collinearity that we can weed out, but we'll have to more to find them.

In [279]:
all_data.to_csv("All_Data.csv", index=False)

In [280]:
all_data

YEAR  CONF              TEAM  SEED  ROUND  K TEMPO   KADJ T    K OFF  \
0     2024   MAC             Akron    14      0  66.7747  65.8933  107.841   
1     2024   SEC           Alabama     4      0  74.1625  72.6461  121.712   
2     2024   P12           Arizona     2      0  73.3760  71.8379  117.653   
3     2024   SEC            Auburn     4      0  70.9629  69.7887  117.364   
4     2024   B12            Baylor     3      0  66.8428  65.6032  117.262   
...    ...   ...               ...   ...    ...      ...      ...      ...   
1074  2008    BE     West Virginia     7     16  66.2223  63.0942  111.246   
1075  2008    SB  Western Kentucky    12     16  68.7887  66.5432  110.914   
1076  2008  BSth          Winthrop    13     64  63.7666  60.9746  101.646   
1077  2008   B10         Wisconsin     3     16  62.1920  60.8809  108.495   
1078  2008   A10            Xavier     3      8  65.1599  62.8978  115.214   

        KADJ O     K DEF  ...  CLOSE TWOS FG%D  CLOSE TWOS D SHARE  \
0     107.0090   99.7096  ...             58.0                37.7   
1     125.6010  108.0310  ...             60.2                37.9   
2     121.1250   96.9086  ...             57.4                30.3   
3     120.5790   95.7924  ...             48.3                42.0   
4     122.4900  104.1660  ...             65.7                29.8   
...        ...       ...  ...              ...                 ...   
1074  113.5700   95.1209  ...              NaN                 NaN   
1075  111.0760   94.3478  ...              NaN                 NaN   
1076   99.6932   92.1731  ...              NaN                 NaN   
1077  113.6190   86.5633  ...              NaN                 NaN   
1078  116.4130   95.7631  ...              NaN                 NaN   

      FARTHER TWOS FG%  FARTHER TWOS SHARE  FARTHER TWOS FG%D  \
0                 43.8                24.6               39.9   
1                 35.7                10.8               37.6   
2                 37.9                23.1               38.5   
3                 41.6                21.1               33.6   
4                 37.5                26.9               40.0   
...                ...                 ...                ...   
1074               NaN                 NaN                NaN   
1075               NaN                 NaN                NaN   
1076               NaN                 NaN                NaN   
1077               NaN                 NaN                NaN   
1078               NaN                 NaN                NaN   

      FARTHER TWOS D SHARE  THREES FG%  THREES SHARE  THREES FG%D  \
0                     25.6        32.0          41.5         30.0   
1                     25.3        36.5          46.8         31.9   
2                     31.5        37.1          32.6         33.4   
3                     24.7        35.2          37.5         29.8   
4                     34.6        39.4          38.3         33.5   
...                    ...         ...           ...          ...   
1074                   NaN         NaN           NaN          NaN   
1075                   NaN         NaN           NaN          NaN   
1076                   NaN         NaN           NaN          NaN   
1077                   NaN         NaN           NaN          NaN   
1078                   NaN         NaN           NaN          NaN   

      THREES D SHARE  
0               36.8  
1               36.7  
2               38.2  
3               33.3  
4               35.5  
...              ...  
1074             NaN  
1075             NaN  
1076             NaN  
1077             NaN  
1078             NaN  

[1079 rows x 90 columns]

# EDA